CMakeLists.txt

```diff
@@ -120,7 +120,7 @@ zephyr_compile_options(
 )
 
 zephyr_compile_options(
-  $<$<COMPILE_LANGUAGE:C>:-std=c99>
+  $<$<COMPILE_LANGUAGE:C>:-std=c11>
```

Comment:
This looks dangerous.

Perhaps migrating from c99 to c11 should be done in a separate and earlier PR, to isolate the change and to make sure it is safe.

Risks:

We might be using deprecated c99 features.
We might be supporting a toolchain that does not support c11.

Reply:
Agree with @SebastianBoe. Variable Length Arrays is one such difference; they were mandatory in C99 but made optional in C11. Zephyr uses VLAs in for red-black tree and mempool. (At Oticon, our LLVM compiler from Synopsys does in fact not support VLAs in either C99 or C11. We are going to write an alloca and send a PR)

Reply:
So, full disclosure: I'm the culprit for the VLAs in both spots. :)

We're actually relying on extensions to standards in many locations in Zephyr, this certainly isn't the only one.  The regression in C11 is unfortunate and shortsighted on the part of the standards committee, but I don't see that it's really relevant to whether or not we pass -std=c11 or not.  GCC and clang, obviously, don't actually disable them when in C11 mode.

Reply:
To be clear: my view is that a -std=xxx argument is a practical platform decision based on what is needed to get the Zephyr source as it exists to build correctly.  The decision as to whether Zephyr is "really" written in C99 or C11 or whatever (or can use VLAs, or GCC statement expressions, or __attribute__((cleanup)), etc..)  is a policy thing that belongs in documentation.  Personally I think this is fine.

Reply:
Just to make a bit more of noise: I agree with @andyross here. In general we should not need to reach a compromise about the compiler switches so they are bearable for all platforms, but just set them per arch or soc and compiler. Seems 2011 is good enough for all platforms (at least all sanitycheck compiles for) so it should be safe to take it in.
If currently it is not easy enough to override from another platform cmake file, then let's make it so.

Reply:
@andyross 
> We're actually relying on extensions to standards in many locations in Zephyr, this certainly isn't the only one

True. Oticon will have to clean that up.

Reply:
Ideally by evolving your compiler, right? :)  I mean, there's room for argument in all of these places, but Zephyr is an open source project, and bending over backward to support oddball proprietary toolchains when they fall behind pervasive industry practice[1] isn't really part of the charter.

[1] I mean, VLA's are hardly rare, and they **are** required by C99, after all!

Reply:
@andyross I definitely agree :) Unfortunately, the reality is that some embedded toolchains (like ours) suffer from half-baked proprietary compilers. Sad but true.

Reply:
We did have c99 set, but it was not enforced. We do need to settle on a standard and enforce it, right now we have a mix. We should be able to support commercial compilers and those compilers which do support one standard but the other. Of course we need to be forward looking and set a minimum. 
The issue with clang is that it has a completely different warnings enabled by default and while Zephyr builds just fine with c99, it generate lots of warning, so we could go and suppress them instead of changing the standard.